neoemf-core/src/main/java/fr/inria/atlanmod/neoemf/bind/ClasspathCollector.java

```diff
@@ -0,0 +1,193 @@
+     */
+    @Nonnull
+    public static ClasspathCollector getInstance() {
+        return Holder.INSTANCE;
+    }
+
+    @Nonnull
+    @Override
+    public Set<URL> get() {
+        Log.debug("Waiting for the classpath analysis... [phase = {0}]", phaser.getRegisteredParties());
+
+        // Waiting for classpath analysis
+        phaser.awaitAdvance(READY);
+        Log.debug("Classpath analysis is complete [phase = {0}]", phaser.getRegisteredParties());
+
+        return registeredUris.stream()
+                .map(URL_TO_URI::revert)
+                .collect(Collectors.toSet());
+    }
+
+    /**
+     * Scans all URLs provided by the given {@code callable} and retain only those that can be related to NeoEMF.
+     *
+     * @param urlCollector the {@link URLCollector} to analyze
+     *
+     * @throws NullPointerException if {@code urlCollector} is {@code null}
+     */
+    public void register(URLCollector urlCollector) {
+        checkNotNull(urlCollector, "urlCollector");
+
+        phaser.register();
+        Log.debug("Starting a classpath analysis... [phase = {0}]", phaser.getRegisteredParties());
+
+        analysisPool.submit(() -> {
+            try {
+                // One configuration for one task
+                ConfigurationBuilder baseConfig = new ConfigurationBuilder()
+                        .setExecutorService(Bindings.getBindingPool())
+                        .setScanners(new TypeAnnotationsScanner(), new SubTypesScanner());
+
+                // Filter URLs, and remove any that cannot be related to NeoEMF (Java, EMF,...)
+                checkNotNull(urlCollector.get(), "urlCollector.get").stream()
+                        .filter(url -> isRelated(url, baseConfig))
+                        .map(URL_TO_URI::convert)
+                        .collect(Collectors.toCollection(() -> registeredUris));
+            }
+            catch (Exception e) {
+                Log.warn(e);
+            }
+            finally {
+                phaser.arriveAndDeregister();
+                Log.debug("Stopping a classpath analysis [phase = {0}]", phaser.getRegisteredParties());
+            }
+        });
+    }
+
+    /**
+     * Returns {@code true} if the {@code url} contains any element related to NeoEMF.
+     *
+     * @param url        the url to scan
+     * @param baseConfig the base configuration of the classpath scanning
+     *
+     * @return {@code true} if the {@code url} contains any element related to NeoEMF
+     */
+    private boolean isRelated(URL url, ConfigurationBuilder baseConfig) {
+        try {
+            return !new Reflections(baseConfig.setUrls(url))
+                    .getTypesAnnotatedWith(BASE_ANNOTATION)
+                    .isEmpty();
+        }
+        catch (RuntimeException e) {
+            return false;
+        }
+    }
+
+    /**
+     * The initialization-on-demand holder of the singleton of this class.
+     */
+    @Static
+    private static final class Holder {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Add a private constructor to hide the implicit public one. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1118)